File: Examples/intel_examples/examples_core_c/datafittingc/source/dfsbinsearch.c
/*******************************************************************************
* Copyright 2010-2020 Intel Corporation.
*
* This software and the related documents are Intel copyrighted  materials,  and
* your use of  them is  governed by the  express license  under which  they were
* provided to you (License).  Unless the License provides otherwise, you may not
* use, modify, copy, publish, distribute,  disclose or transmit this software or
* the related documents without Intel's prior written permission.
*
* This software and the related documents  are provided as  is,  with no express
* or implied  warranties,  other  than those  that are  expressly stated  in the
* License.
*******************************************************************************/

/*
!  Content:
!    Cells search in blocks Example Program Text
!******************************************************************************/

#include <stdio.h>

#include "mkl.h"
#include "errcheck.inc"
#include "generatedata.inc"
#include "rescheck.inc"

#define N                     10    // number of breakpoints
#define NSITE_BLOCK           15    // number of sites for cell search in block
#define NBLOCKS                3    // number of blocks
#define LDATA                  5    // length of the datahint array

#define NSITE (NBLOCKS*NSITE_BLOCK) // total number of sites for cell search

#define LEFT_LIMIT         -10.0    // left  limit of search interval
#define RIGHT_LIMIT         10.0    // right limit of search interval

#define MEAN                 0.0    // mean of the sites distribution
#define SIGMA                1.0    // sandard deviation of the sites
                                    // distribution

int main()
{
    DFTaskPtr task;                 // Data Fitting task descriptor
    MKL_INT nx;                     // number of break points
    MKL_INT xhint;                  // additional info about break points
    MKL_INT nsite_bl;               // number of search sites in one block
    MKL_INT nsite;                  // total number of sites
    MKL_INT nblocks;                // number of blocks of sites
    MKL_INT sitehint;               // search sites storage format
    MKL_INT cell[NSITE];            // indices of cells containing
    MKL_INT *cell_ptr;              // search sites
    float x[N];                     // array of break points
    float site[NSITE];              // array of search sites
    float *site_ptr;
    float datahint[LDATA];          // additional info about the structure
                                    // of search sites
    float a, sigma;                 // parameters of the search sites
                                    // distribution
    float left = LEFT_LIMIT;        // left  limit of the search interval
    float right = RIGHT_LIMIT;      // right limit of the search interval

    MKL_INT test_cell[NSITE];
    int i, j, errcode = 0;
    int errnums = 0;
    float r;

    /***** Initializing parameters for Data Fitting task *****/

    /***** Parameters describing search interval *****/
    nx        = N;
    xhint     = DF_NON_UNIFORM_PARTITION;

    /***** Parameter describing number of search sites *****/
    nsite_bl  = NSITE_BLOCK;
    nsite     = NSITE;
    /* No additional info is provided in case of cell search */
    sitehint  = DF_NO_HINT;

    /***** Generate uniformly distributed search sites *****/
    errcode = sUniformRandSortedData( x, left, right, nx );
    CheckDfError(errcode);

    /***** Parameter describing additional info about the search sites *****/
    /* Presumably most likely cell is cell number (N/2 + 1) */
    datahint[0] = 1.0;
    datahint[1] = (float)DF_APRIORI_MOST_LIKELY_CELL;
    datahint[2] = 0.0;
    datahint[3] = 1.0;
    datahint[4] = (float)((nx/2) + 1);

    /***** Generate normally distributed break points *****/
    a         = MEAN;
    sigma     = SIGMA;
    errcode = sPeakData( site, a, sigma, nsite );
    CheckDfError(errcode);

    /***** Create Data Fitting task *****/
    errcode = dfsNewTask1D( &task, nx, x, xhint, 0, 0, 0 );
    CheckDfError(errcode);

    /***** Perform cells search *****/

    nblocks = NBLOCKS;

    site_ptr = site;
    cell_ptr = cell;
    for ( i = 0; i < nblocks; i++ )
    {
        errcode = dfsSearchCells1D( task, DF_METHOD_STD, nsite_bl, site_ptr,
                                    sitehint, datahint, cell_ptr );
        CheckDfError(errcode);

        site_ptr += nsite_bl;
        cell_ptr += nsite_bl;
    }

    /***** Delete Data Fitting task *****/
    errcode = dfDeleteTask( &task );
    CheckDfError(errcode);

    /***** Check search results *****/
    errcode = sFindCells( nx, x, nsite, site, test_cell );

    for ( i = 0; i < nsite; i++ )
    {
        if ( test_cell[i] != cell[i] ) errnums++;
    }

    /***** Print results *****/
    printf("Number of break points : %d\n", (int)nx);
    printf("Number of search sites : %d\n", (int)nsite);

    /***** Print given function *****/
    printf("\n  X:\n");

    for( j = 0; j < nx; j++ )
        printf(" %+lf\n", x[j]);

    printf("\nResults of cell search:\n");
    printf("    Site          Computed idx      Expected idx\n");
    for ( i = 0; i < nsite; i++ )
    {
        printf(" %+11.6lf   %15d   %15d\n", site[i], (int)(cell[i]),
            (int)(test_cell[i]));
    }

    /***** Print summary of the test *****/
    if (errnums != 0)
    {
        printf("\n\nError: Computed cell indices are incorrect\n");
        return 1;
    }
    else
    {
        printf("\n\nComputed cell indices are correct\n");
    }

    return 0;
}

File: Examples/intel_examples/examples_core_f/datafittingf/source/dfsbinsearch.f
!===============================================================================
! Copyright 2010-2020 Intel Corporation.
!
! This software and the related documents are Intel copyrighted  materials,  and
! your use of  them is  governed by the  express license  under which  they were
! provided to you (License).  Unless the License provides otherwise, you may not
! use, modify, copy, publish, distribute,  disclose or transmit this software or
! the related documents without Intel's prior written permission.
!
! This software and the related documents  are provided as  is,  with no express
! or implied  warranties,  other  than those  that are  expressly stated  in the
! License.
!===============================================================================

!  Content:
!    Cells search in blocks Example Program Text
!*******************************************************************************

      include 'mkl_df.f90'
      include "errcheck.inc"
      include "rescheck.inc"
      include "generatedata.inc"

      PROGRAM MKL_DF_TEST

      USE MKL_DF_TYPE
      USE MKL_DF
      USE DF_GENERATE_DATA
      USE DF_EXAMPLE_RESCHECK


      ! number of breakpoints
      INTEGER,PARAMETER :: N           = 10
      ! number of sites for cell search in block
      INTEGER,PARAMETER :: NSITE_BLOCK = 15
      ! number of blocks
      INTEGER,PARAMETER :: NNBLOCKS    = 3
      ! length of the datahint array
      INTEGER,PARAMETER :: LDATA       = 5
      ! total number of sites for cell search
      INTEGER,PARAMETER :: NNSITE      = (NNBLOCKS*NSITE_BLOCK)
      ! left  limit of search interval
      REAL(4),PARAMETER :: LEFT_LIMIT  = -10.0
      ! right limit of search interval
      REAL(4),PARAMETER :: RIGHT_LIMIT = 10.0

      ! mean of the sites distribution
      REAL(4),PARAMETER :: MEAN_  = 0.0
      ! sandard deviation of the sites distribution
      REAL(4),PARAMETER :: SIGMA_ = 1.0

      ! Data Fitting task descriptor
      TYPE (DF_TASK) task

      ! number of break points
      INTEGER :: nx
      ! additional info about break points
      INTEGER :: xhint
      ! number of search sites in one block
      INTEGER :: nsite_bl
      ! total number of sites
      INTEGER :: nsite
      ! number of blocks of sites
      INTEGER :: nblocks
      ! search sites storage format
      INTEGER :: sitehint
      ! indices of cells containing search sites
      INTEGER,TARGET  :: cell(NNSITE)
      INTEGER,POINTER :: cell_ptr(:)
      ! array of break points
      REAL(4) :: x(N)
      ! array of search sites
      REAL(4),TARGET :: site(NNSITE)
      REAL(4),POINTER :: site_ptr(:)
      ! additional info about the structure of search sites
      REAL(4) :: datahint(LDATA)
      ! parameters of the search sites distribution
      REAL(4) :: a,sigma
      ! limits of the search interval
      REAL(4) :: left, right
      ! method that is used to perform calculations
      INTEGER :: method

      INTEGER :: test_cell(NNSITE)
      INTEGER :: i,j
      INTEGER :: errnums
      INTEGER(4) :: errcode
      REAL(4) :: r

      !***** Initializing parameters for Data Fitting task *****

      errcode = 0
      errnums = 0

      !***** Parameters describing search interval *****
      nx = N
      xhint = DF_NON_UNIFORM_PARTITION

      !***** Parameter describing number of search sites *****
      nsite_bl  = NSITE_BLOCK
      nsite     = NNSITE
      !  No additional info is provided in case of cell search
      sitehint  = DF_NO_HINT

      !***** Generate uniformly distributed search sites *****
      left = LEFT_LIMIT
      right = RIGHT_LIMIT
      errcode = sUniformRandSortedData( x, left, right, nx )
      CALL CheckDfError(errcode)

      !***** Parameter describing additional info about the search sites *****
      !  Presumably most likely cell is cell number (N/2 + 1)
      datahint(1) = 1.0
      datahint(2) = FLOAT(DF_APRIORI_MOST_LIKELY_CELL)
      datahint(3) = 0.0
      datahint(4) = 1.0
      datahint(5) = FLOAT(((nx/2) + 1))

      !***** Generate normally distributed break points *****
      a         = MEAN_;
      sigma     = SIGMA_;
      errcode = sPeakData( site, a, sigma, nsite )
      CALL CheckDfError(errcode)

      !***** Create Data Fitting task *****
      errcode = dfsNewTask1D( task, nx, x, xhint )
      CALL CheckDfError(errcode)

      !***** Perform cells search *****
      nblocks = NNBLOCKS

      method = DF_METHOD_STD

      DO i = 1, nblocks
        site_ptr => site((i-1)*nsite_bl+1:)
        cell_ptr => cell((i-1)*nsite_bl+1:)
        errcode = dfsSearchCells1D( task, method, nsite_bl,              &
     &     site_ptr, sitehint, datahint, cell_ptr )
        CALL CheckDfError(errcode)
      END DO

      !***** Check search results *****
      errcode = sFindCells( nx, x, nsite, site, test_cell )
      DO i = 1, nsite
        test_cell(i) = test_cell(i) - 1
        IF ( test_cell(i) /= cell(i) ) errnums = errnums+1
      END DO

      !***** Print given function *****
      WRITE (*,901) "Number of break points : ",nx
      WRITE (*,901) "Number of search sites : ",nsite
      WRITE (*,902) "  X:"
      DO j = 1, nx
        WRITE (*,903) " ",x(j)
      END DO
      WRITE (*,902) "Results of cell search:"
      WRITE (*,904) "    Site          Computed idx      Expected idx"
      DO i = 1, nsite
        WRITE (*,905) " ",site(i),"   ",cell(i),"   ",test_cell(i)
      END DO

      !***** Delete Data Fitting task *****
      errcode = dfDeleteTask( task )
      CALL CheckDfError(errcode)

      !***** Print summary of the test *****
      IF (errnums /= 0) THEN
        WRITE (*,906) "","Error: Computed cell indices are incorrect"
        STOP 1
      ELSE
        WRITE (*,906) "","Computed cell indices are correct"
      END IF
      STOP 0

901   FORMAT (A,I0)
902   FORMAT (/A)
903   FORMAT (A,SP,F11.6)
904   FORMAT (99A)
905   FORMAT (A,SP,F11.6,A,I15,A,I15)
906   FORMAT (//A)
      END PROGRAM
